cmd/cli/config/set.go

```diff
@@ -26,40 +27,58 @@ func newSetCmd() *cobra.Command {
 			if err != nil {
-				return fmt.Errorf("failed to setup repo: %w", err)
+				return fmt.Errorf("failed to get config flag: %w", err)
 			}
-			return setConfig(cmd.PersistentFlags().Lookup("config").Value.String(), args[0], args[1:]...)
-		},
-		// provide auto completion for arguments to the `set` command
-		ValidArgsFunction: setAutoComplete,
-	}
 
-	bacalhauCfgDir := "bacalhau"
-	bacalhauCfgFile := config.DefaultFileName
+			// If --config flag is set, use it to set the Viper key
+			if configFlag != "" {
+				viper.Set(cliflags.RootCommandConfigFiles, []string{configFlag})
+			}
 
-	usrCfgDir, err := os.UserConfigDir()
-	if err != nil {
-		log.Warn().Err(err).Msg("Failed to find user-specific configuration directory. Using current directory to write config.")
-	} else {
-		bacalhauCfgDir = filepath.Join(usrCfgDir, bacalhauCfgDir)
-		if err := os.MkdirAll(bacalhauCfgDir, util2.OS_USER_RWX); err != nil {
-			// This means we failed to create a directory either in the current directory, or the user config dir
-			// indicating a some-what serious misconfiguration of the system. We panic here to provide as much
-			// detail as possible.
-			log.Panic().Err(err).Msgf("Failed to create bacalhau configuration directory: %s", bacalhauCfgDir)
-		}
-		bacalhauCfgFile = filepath.Join(bacalhauCfgDir, bacalhauCfgFile)
+			var configPath string
+			// load configs to get the config file path
+			rawConfig, err := util.SetupConfigType(cmd)
+			if err != nil {
+				return fmt.Errorf("failed to setup config: %w", err)
+			}
+
+			configPath = rawConfig.ConfigFileUsed()
+			if configPath == "" {
+				// we fall back to the default config file path $BACALHAU_DIR/config.yaml
+				// this requires initializing a new or opening an existing data-dir
+				bacalhauConfig, err := util.UnmarshalBacalhauConfig(rawConfig)
+				if err != nil {
+					return fmt.Errorf("failed to decode bacalhau config: %w", err)
+				}
+				_, err = util.SetupRepo(bacalhauConfig)
+				if err != nil {
+					return fmt.Errorf("failed to setup repo: %w", err)
+				}
+				configPath = filepath.Join(bacalhauConfig.DataDir, config.DefaultFileName)
+
+				// create the config file if it doesn't exist
+				if _, err := os.Stat(configPath); os.IsNotExist(err) {
+					if err := os.WriteFile(configPath, []byte{}, util2.OS_USER_RWX); err != nil {
+						return fmt.Errorf("failed to create default config file %s: %w", configPath, err)
+					}
+				}
+			}
+
+			return setConfig(configPath, args[0], args[1:]...)
+		},
+		// Provide auto completion for arguments to the `set` command
+		ValidArgsFunction: setAutoComplete,
 	}
 
-	setCmd.PersistentFlags().String("config", bacalhauCfgFile, "Path to the config file")
+	setCmd.PersistentFlags().String("config", "", "Path to the config file (default is $BACALHAU_DIR/config.yaml)")
 	return setCmd
 }
 
 func setConfig(cfgFilePath, key string, value ...string) error {
+	log.Info().Msgf("Writing config to %s", cfgFilePath)
```

Comment:
@wdbaruni on line 94 of this method we write the config file, this call ensures a file is create if one doesn't exist.

Reply:
I see what you mean. Removed in https://github.com/bacalhau-project/bacalhau/pull/4466/commits/f43063212522c9a2616e028915ee767490fc618b